### ISSUE TITLE
Require notification permission for actions to control media

### ISSUE THREAD

**Issue by logixoul** [keymapperorg/KeyMapper#1353]
**Developer TODO (don't remove)**
- [ ] write tests. put issue number in comment
- [ ] update documentation

**Discord message link/email recipient**

**Describe the bug**
Can't control app that is not visible (i.e. on top).

**To Reproduce**
1. Connect an external usb keyboard to your phone.
2. In Key Mapper, bind that keyboard's "space" button to the "Play/Pause media playback" action.
3. Open a media player app (Musicolet or Smart AudioBook Player).
4. Inside that app, begin playback of any audio file.
5. Press "space" - it works fine here.
6. Now press the Android "home" button to go to the home screen.
7. Press "space" - nothing happens.

**Expected behavior**
The "space" key should've paused the playback.

**Smartphone (please complete the following information):**
 - Device: Samsung Galaxy A15 4G
 - Android version: Android 14

**Additional context**
This works fine in a competitor app called "Button mapper".
[key_mapper_report_20241205-114454.zip](https://github.com/user-attachments/files/18021177/key_mapper_report_20241205-114454.zip)
[mappings_20241205-115116.zip](https://github.com/user-attachments/files/18021178/mappings_20241205-115116.zip)

**Comment by sds100** [comment]
And if you use the action for play/pause media for a specific app?

**Comment by logixoul** [comment]
@sds100 I now tried that. I'll try to describe the exact sequence of what happened:
1. I configured a trigger (key "z") to perform the "play/pause media for a specific app" action.
2. THEN, for the first time, Key Mapper asked me for the notification-access permission. I allowed it.
3. I opened "Smart AudioBook Player", minimized it using the Home button and tried the "z" key. It worked!
4. Then I tried my original shortcut (the "space" key, still mapped to plain "play/pause media") and IT WORKED TOO!

Conclusion: It seems that, once KeyMapper has asked for notif permissions and I have allowed them, the bug described in my original post goes away.

A probable fix would be to have KeyMapper ask for notif permissions right when the user adds a plain "play/pause media" action.

Thank you!! :)

**Comment by sds100** [comment]
Ah, I didn't realise it needs notification permission. I will add that to those actions. Thanks for figuring it out!

**Comment by sds100** [comment]
I can't really reproduce this on the stock Android 14 emulator or my Android 14 Pixel 7. Do you have any other Android 14+ devices lying around? Perhaps this is a Samsung only thing for some reason.

**Comment by logixoul** [comment]
Sorry, I don't. My other devices are:
- Moto G62 5G (android 13)
- Samsung Galaxy A14 (android 13)
- Samsung Galaxy A12 (android 12)

Anything else I can do to help?

**Comment by sds100** [comment]
Do you see the same behavior with Spotify, YouTube or Youtube Music? I tested those.

**Comment by logixoul** [comment]
@sds100 
To test this, I removed the "Notification Access" permission for Key Mapper (to hopefully bring things to the "original state"), then tried to reproduce the bug but the bug did NOT happen - neither for YouTube Music nor for Smart AudioBook Player.

I also removed the "Notification Access" permission for Button Mapper (BM), created a BM mapping (from key '5' to 'play/pause media') (BM did NOT ask me to re-grant the permission) and tried to reproduce the bug but the bug did NOT happen - the '5' key successfully paused the audio playing from my "minimized" ("background") YT Music. Same with Smart AudioBook Player

Anything else I can help with?

TLDR: I'm not able to reproduce the bug anymore, no matter what I do.

**Comment by sds100** [comment]
Interesting, thank you for the thorough investigation 😅. Not sure it is a bug with Key Mapper then. I will close this for now and if it resurfaces or if you can figure out how to reproduce it then we will try to fix it. You're the first person to mention this issue.
